<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="utf-8"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<meta content="None" name="description"/>
<link href="img/favicon.ico" rel="shortcut icon"/>
<title>Caprion</title>
<link href="css/bootstrap.min.css" rel="stylesheet"/>
<link href="css/font-awesome.min.css" rel="stylesheet"/>
<link href="css/base.css" rel="stylesheet"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet"/>
<script defer="" src="js/jquery-1.10.2.min.js"></script>
<script defer="" src="js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Bash.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/HTML.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Java.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/JavaScript.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Perl.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Python.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/R.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/rust.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/yaml.min.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
</head>
<body class="homepage">
<div class="navbar fixed-top navbar-expand-lg navbar-dark bg-dark">
<div class="container">
<a class="navbar-brand" href=".">Caprion</a>
<!-- Expander button -->
<button class="navbar-toggler" data-target="#navbar-collapse" data-toggle="collapse" type="button">
<span class="navbar-toggler-icon"></span>
</button>
<!-- Expanded navigation -->
<div class="navbar-collapse collapse" id="navbar-collapse">
<!-- Main navigation -->
<ul class="nav navbar-nav">
<li class="navitem">
<a class="nav-link" href="."></a>
</li>
<li class="navitem active">
<a class="nav-link" href=".">Analysis</a>
</li>
<li class="navitem">
<a class="nav-link" href="pilot/">Pilot studies</a>
</li>
<li class="navitem">
<a class="nav-link" href="pilot/gwas2/">gwas2</a>
</li>
<li class="navitem">
<a class="nav-link" href="pilot/autoencoder/">Autoencoder</a>
</li>
<li class="navitem">
<a class="nav-link" href="pilot/Notes/">Notes</a>
</li>
</ul>
<ul class="nav navbar-nav ml-auto">
<li class="nav-item">
<a class="nav-link" data-target="#mkdocs_search_modal" data-toggle="modal" href="#">
<i class="fa fa-search"></i> Search
                            </a>
</li>
<li class="nav-item">
<a class="nav-link" href="." rel="prev">
<i class="fa fa-arrow-left"></i> Previous
                                </a>
</li>
<li class="nav-item">
<a class="nav-link" href="pilot/" rel="next">
                                    Next <i class="fa fa-arrow-right"></i>
</a>
</li>
<li class="nav-item">
<a class="nav-link" href="https://github.com/jinghuazhao/Caprion"><i class="fa fa-github"></i> GitHub</a>
</li>
</ul>
</div>
</div>
</div>
<div class="container">
<div class="row">
<div class="col-md-3"><div class="navbar-light navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
<div class="navbar-header">
<button class="navbar-toggler collapsed" data-target="#toc-collapse" data-toggle="collapse" title="Table of Contents" type="button">
<span class="fa fa-angle-down"></span>
</button>
</div>
<div class="navbar-collapse collapse card bg-secondary" id="toc-collapse">
<ul class="nav flex-column">
<li class="nav-item" data-level="1"><a class="nav-link" href="#analysis">Analysis</a>
<ul class="nav flex-column">
<li class="nav-item" data-level="2"><a class="nav-link" href="#workflow-experimental">workflow (experimental)</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#programs1">Programs1</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#1-data-handling-and-pca-projection">1. Data handling and PCA projection</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#2-ggm">2. GGM</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#3-wgcna">3. WGCNA</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#4-pca-and-clustering">4. PCA and clustering</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#5-pgwas2">5. pGWAS2</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#6-meta-analysis">6. Meta-analysis</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#7-variant-identification">7. Variant identification</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#8-hla-imputation4">8. HLA imputation4</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#9-lookup">9. Lookup</a>
<ul class="nav flex-column">
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div></div>
<div class="col-md-9" role="main">
<h1 id="analysis">Analysis</h1>
<h2 id="workflow-experimental">workflow (experimental)</h2>
<pre><code class="language-bash">module add ceuadmin/snakemake

snakemake -s workflow/rules/cojo.smk -j1
snakemake -s workflow/rules/report.smk -j1
snakemake -s workflow/rules/cojo.smk -c --profile workflow
</code></pre>
<p>and use <code>--unlock</code> when necessary.</p>
<h2 id="programs1">Programs<sup id="fnref:directory"><a class="footnote-ref" href="#fn:directory">1</a></sup></h2>
<p>Earlier work was done in a named sequence.</p>
<pre><code>1_pca_projection.sh
2_ggm.R
3_wgcna.R
4_pca_clustering.R
5_pgwas.sh
6_meta_analysis.sh
7_merge.sh
8_hla.sh
9_lookup.sh
</code></pre>
<h2 id="1-data-handling-and-pca-projection">1. Data handling and PCA projection</h2>
<p>The pipeline follows HGI contributions nevertheless only serves for reassurance since the study samples were carefully selected. </p>
<h2 id="2-ggm">2. GGM</h2>
<p>The results are ready to report.</p>
<h2 id="3-wgcna">3. WGCNA</h2>
<p>This can be finalised according to the Science paper.</p>
<h2 id="4-pca-and-clustering">4. PCA and clustering</h2>
<p>The groupings based on proteins can be made on three phases altogether and instead of a classification indicator the first three PCs are used.</p>
<p>The PLINK2 has been consistent in the pilot studies, so <code>scale()</code> operaions can be used in the inverse normal transformations.</p>
<p>The phenotypic data is generated in accordance with the double transformations as in SCALLOP-Seq analysis.</p>
<h2 id="5-pgwas2">5. pGWAS<sup id="fnref:pGWAS"><a class="footnote-ref" href="#fn:pGWAS">2</a></sup></h2>
<p>The bgen files were extracted from a list of all samples, the variant IDs of which were replaced when RSid is missing (.).</p>
<p>The bgen generation is moved into .sb based on cclake but can be switched back to cardio by uncommenting the ##SBATCH lines.</p>
<p>The (sb)atch file is extended to produce Q-Q/Manhattan/LocusZoom plots and extreme p values are possible for all plots. Note that LocusZoom 1.4 does not contain 1000Genomes build 37 genotypes for chromosome X and therefore they are supplemented with local files in the required format, namely, <code>locuszoom_1.4/data/1000G/genotypes/2014-10-14/EUR/chrX.[bed, bim, fan]</code>.</p>
<h2 id="6-meta-analysis">6. Meta-analysis</h2>
<p>This follows from the SCALLOP/INF implementation, as designed analogous to a Makefile, i.e.,</p>
<pre><code class="language-bash">6_meta_analysis &lt;task&gt;
</code></pre>
<p>where task = METAL_list, METAL_files, METAL_analysis, respectively in sequence.</p>
<p>To extract significant variants one may resort to <code>awk 'NR==1||$12&lt;log(1e-6)/log(10)' 1433B-1.tbl</code>, say.</p>
<h2 id="7-variant-identification">7. Variant identification</h2>
<p>An iterative merging scheme is employed; the HLA region is simplified but will be specifically handled. Somewhat paradoxically, forest plots are also obtained here<sup id="fnref:metal"><a class="footnote-ref" href="#fn:metal">3</a></sup>.</p>
<h2 id="8-hla-imputation4">8. HLA imputation<sup id="fnref:HLA"><a class="footnote-ref" href="#fn:HLA">4</a></sup></h2>
<p>This is experimented on several software including HIBAG, CookHLA and SNP2HLA as desribed <a href="https://cambridge-ceu.github.io/csd3/applications/CookHLA.html">here</a>. The whole cohort imputation requests resources exceeding the system limits, so a cardio SLURM job is used instead.</p>
<p>The hped file from CookHLA (or converted from HIBAG) can be used by HATK for association analysis while the advantage of SNP2HLA is that binary ped files are ready for use as usual.</p>
<h2 id="9-lookup">9. Lookup</h2>
<div class="footnote">
<hr/>
<ol>
<li id="fn:directory">
<p>Directories</p>
<p>This is per Caprion project</p>
<p><code>module load miniconda3/4.5.1
export csd3path=/rds/project/jmmh2/rds-jmmh2-projects/olink_proteomics/scallop/miniconda37
source ${csd3path}/bin/activate</code></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>pgwas</td>
<td>pGWAS</td>
</tr>
<tr>
<td>METAL</td>
<td>Meta-analysis</td>
</tr>
<tr>
<td>HLA</td>
<td>HLA imputation</td>
</tr>
<tr>
<td>reports</td>
<td>Reports</td>
</tr>
</tbody>
</table>
<p><a class="footnote-backref" href="#fnref:directory" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:pGWAS">
<p>Protein GWAS</p>
<p>GCTA/fastGWA employs MAF&gt;=0.0001 (~56%) and geno=0.1 so potentially we can have .bgen files as such to speed up.</p>
<p>GCTA uses headerless phenotype files, so <strong>the following section from <code>5_pgwas.sh</code> is run</strong> in preparation.</p>
<p><code>sed -i '1d' ${caprion}/work/caprion-1.pheno
sed -i '1d' ${caprion}/work/caprion-2.pheno
sed -i '1d' ${caprion}/work/caprion-3.pheno</code></p>
<p>at <code>pilot/work</code> while the original version is saved at <code>analysis/work/</code>.</p>
<p>It looked to take 3.5 days on Cardio without unfiltered genotypes but ~12 hours on cclake, and once these are taken care of the analysis can be propagated. <a class="footnote-backref" href="#fnref:pGWAS" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:metal">
<p>incomplete gamma function</p>
<p>The .info files for proteins BROX and CT027 could not be obtained from METAL 2020-05-05 with the following error message,</p>
<p><code>FATAL ERROR -
a too large, ITMAX too small in gamma countinued fraction (gcf)</code></p>
<p>An attempt was made to fix this and reported as a fixable issue to METAL GitHub respository (<a href="https://github.com/statgen/METAL/issues/24"><code>https://github.com/statgen/METAL/issues/24</code></a>).
This has enabled Forest plots for the associate pQTLs. <a class="footnote-backref" href="#fnref:metal" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
<li id="fn:HLA">
<p>Whole cohort imputation is feasible with HIBAG which contains inclusive lists of SNPs based on samples of the following sizes,</p>
<table>
<thead>
<tr>
<th><strong>Locus</strong></th>
<th>A</th>
<th>B</th>
<th>C</th>
<th>DQA1</th>
<th>DQB1</th>
<th>DPB1</th>
<th>DRB1</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>N</strong></td>
<td>1857</td>
<td>2572</td>
<td>1866</td>
<td>1740</td>
<td>1924</td>
<td>1624</td>
<td>2436</td>
</tr>
</tbody>
</table>
<p>while the reference panel is based on the 1000Genomes data (N=503) with SNP2HLA and CookHLA.</p>
<p>It is of note that <code>1000G_REF.EUR.chr6.hg18.29mb-34mb.inT1DGC.markers</code> in the 1000Genomes reference panel has 465 variants with HLA prefix and the partition is as follows,</p>
<table>
<thead>
<tr>
<th><strong>Locus</strong></th>
<th>A</th>
<th>B</th>
<th>C</th>
<th>DQA1</th>
<th>DQB1</th>
<th>DPB1</th>
<th>DRB1</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>HLA_</strong></td>
<td>98</td>
<td>183</td>
<td>69</td>
<td>0</td>
<td>33</td>
<td>0</td>
<td>82</td>
</tr>
</tbody>
</table>
<p>A recent update: PGG.HLA, <a href="https://www.pggmhc.org/pggmhc/">https://www.pggmhc.org/pggmhc/</a> <a class="footnote-backref" href="#fnref:HLA" title="Jump back to footnote 4 in the text">↩</a></p>
</li>
</ol>
</div></div>
</div>
</div>
<footer class="col-md-12">
<hr/>
<p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
</footer>
<script>
            var base_url = ".",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
<script defer="" src="js/base.js"></script>
<script defer="" src="javascripts/config.js"></script>
<script defer="" src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script defer="" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script defer="" src="search/main.js"></script>
<div aria-hidden="true" aria-labelledby="searchModalLabel" class="modal" id="mkdocs_search_modal" role="dialog" tabindex="-1">
<div class="modal-dialog modal-lg">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="searchModalLabel">Search</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
<form>
<div class="form-group">
<input class="form-control" id="mkdocs-search-query" placeholder="Search..." title="Type search term here" type="search"/>
</div>
</form>
<div id="mkdocs-search-results"></div>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div><div aria-hidden="true" aria-labelledby="keyboardModalLabel" class="modal" id="mkdocs_keyboard_modal" role="dialog" tabindex="-1">
<div class="modal-dialog">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<table class="table">
<thead>
<tr>
<th style="width: 20%;">Keys</th>
<th>Action</th>
</tr>
</thead>
<tbody>
<tr>
<td class="help shortcut"><kbd>?</kbd></td>
<td>Open this help</td>
</tr>
<tr>
<td class="next shortcut"><kbd>n</kbd></td>
<td>Next page</td>
</tr>
<tr>
<td class="prev shortcut"><kbd>p</kbd></td>
<td>Previous page</td>
</tr>
<tr>
<td class="search shortcut"><kbd>s</kbd></td>
<td>Search</td>
</tr>
</tbody>
</table>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div>
</body>
</html>
<!--
MkDocs version : 1.1.2
Build Date UTC : 2023-08-30 07:17:29.985589+00:00
-->
