<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="utf-8"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<meta content="None" name="description"/>
<link href="img/favicon.ico" rel="shortcut icon"/>
<title>Caprion</title>
<link href="css/bootstrap.min.css" rel="stylesheet"/>
<link href="css/font-awesome.min.css" rel="stylesheet"/>
<link href="css/base.css" rel="stylesheet"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet"/>
<script defer="" src="js/jquery-1.10.2.min.js"></script>
<script defer="" src="js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Bash.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/HTML.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Java.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/JavaScript.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Perl.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Python.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/R.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/rust.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/yaml.min.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
</head>
<body class="homepage">
<div class="navbar fixed-top navbar-expand-lg navbar-dark bg-dark">
<div class="container">
<a class="navbar-brand" href=".">Caprion</a>
<!-- Expanded navigation -->
<div class="navbar-collapse collapse" id="navbar-collapse">
<ul class="nav navbar-nav ml-auto">
<li class="nav-item">
<a class="nav-link" data-target="#mkdocs_search_modal" data-toggle="modal" href="#">
<i class="fa fa-search"></i> Search
                            </a>
</li>
</ul>
</div>
</div>
</div>
<div class="container">
<div class="row">
<div class="col-md-3"><div class="navbar-light navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
<div class="navbar-header">
<button class="navbar-toggler collapsed" data-target="#toc-collapse" data-toggle="collapse" title="Table of Contents" type="button">
<span class="fa fa-angle-down"></span>
</button>
</div>
<div class="navbar-collapse collapse card bg-secondary" id="toc-collapse">
<ul class="nav flex-column">
<li class="nav-item" data-level="1"><a class="nav-link" href="#caprion-work">Caprion work</a>
<ul class="nav flex-column">
<li class="nav-item" data-level="2"><a class="nav-link" href="#site-map">Site map</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#coding">Coding</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#gwas2">gwas2</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#autoencoder-ae-vs-pca">autoencoder (AE) vs PCA</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#references">REFERENCES</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#urls">URLs</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#eth-zurich-u-toronto-team-develops-tool-for-bottom-up-proteomics-proteoform-analysis">ETH Zurich, U Toronto Team Develops Tool for Bottom-Up Proteomics Proteoform Analysis</a>
<ul class="nav flex-column">
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div></div>
<div class="col-md-9" role="main">
<h1 id="caprion-work">Caprion work</h1>
<h2 id="site-map">Site map</h2>
<ul>
<li>Pilot (N=196)</li>
<li>data/ contains genotype files in .bgen format</li>
<li>bgen/ PLINK2 results according to .bgen files; summary outputs and sentinels/ directory are in the following directories<ul>
<li>1e-5</li>
<li>1e-6</li>
<li>5e-8</li>
</ul>
</li>
<li>Batch 2 (N=1,488)</li>
<li>data2/ contains genotype files in .bgen format</li>
<li>bgen2/ PLINK2 results according to .bgen files; summary outputs and sentinels/ directory are in the following directories<ul>
<li>1e-5</li>
<li>5e-8</li>
</ul>
</li>
<li>Comparison of pilot and batch 2<ul>
<li>miamiplot</li>
</ul>
</li>
<li>Batch 3 data (N=815)</li>
<li>data3/ .bgen data</li>
<li>bgen3/ PLINK2 results<ul>
<li>1e-5</li>
<li>5e-8</li>
</ul>
</li>
</ul>
<h2 id="coding">Coding</h2>
<ul>
<li>Pilot</li>
<li><a href="caprion.R">caprion.R</a> and <a href="caprion.ini">caprion.ini</a> are for data processing. Their derivatives are in the utils/ subdirectory:<ul>
<li><a href="utils/affymetrix.sh">utils/affymetrix.sh</a> is for variant-specific association analysis.</li>
<li><a href="utils/qctool.sb">utils/qctool.sb</a> is used to extract available sample and genotypes.</li>
<li><a href="utils/qctool.sh">utils/qctool.sh</a> further extracts genotypes with MAF 0.01 only.</li>
<li><a href="utils/plink2.sh">utils/plink2.sh</a> non-SLURM version of association analysis.</li>
<li><a href="utils/qqman.sh">utils/qqman.sh</a> and <a href="utils/qqman.R">utils/qqman.R</a> produce QQ and Manhattan plots.</li>
<li><a href="utils/sentinels_nold.sh">utils/sentinels_nold.sh</a> and <a href="utils/merge.sh">utils/merge.sh</a> select sentinels.</li>
<li><a href="utils/ps.sh">utils/ps.sh</a> and <a href="utils/ps.R">utils/ps.R</a> run through PhenoScanner.</li>
<li><a href="utils/lookup.sh">utils/lookup.sh</a> looks up for overlap with SomaLogic and Olink.</li>
<li><a href="caprion.ipynb">caprion.ipynb</a> is a Jupyter notebook with some preprocessing done by <a href="utils/tensorqtl.sh">utils/tensorqtl.sh</a>.</li>
</ul>
</li>
<li>Batch 2</li>
</ul>
<div class="mermaid">graph TB;
utils/tensoqtl.sh
utils/2020.sh --&gt; EPCR-PROC
utils/2020.sh --&gt; data2/affymetrix.id
data2/affymetrix.id  --&gt; data2/bgen
utils/qctool.sb --&gt; data2/bgen
data2/bgen --&gt; utils/plink2.sh
utils/plink2.sh --&gt; utils/sentinels_nold.sh
utils/sentinels_nold.sh --&gt; merge.sh
</div>
<ul>
<li>Batch 3</li>
</ul>
<div class="mermaid">graph TB
utils/2021.sh
</div>
<h2 id="gwas2">gwas2</h2>
<p>This is a promising alternative showing through RCN3/FCGRN with <a href="utils/gwas2.sh">gwas2.sh</a>.</p>
<div class="mermaid">graph TB;
utils/gwas2.sh --&gt; utils/gwas.do
utils/gwas2.sh --&gt; utils/gwas2.do
</div>
<p>where <code>gwas.do</code> (<code>caprion.dat</code> also contains _invn data) and <code>gwas2.do</code> (<code>gwas2_invn.do</code> for _invn data) are for the pilot and batch 2 data, respectively.</p>
<hr/>
<h2 id="autoencoder-ae-vs-pca">autoencoder (AE) vs PCA</h2>
<p>As shown at <a href="https://www.r-bloggers.com/pca-vs-autoencoders-for-dimensionality-reduction/">R-bloggers<img align="right" height="400" src="https://i0.wp.com/gradientdescending.com/wp-content/uploads/2018/07/reconstruction-1.png" width="560"/></a>,</p>
<blockquote>
<p>autoencoder is better at reconstructing the original data set than PCA when k is small, 
however the error converges as k increases. For very large data sets this difference will be
larger and means a smaller data set could be used for the same error as PCA. When dealing 
with big data this is an important property`.</p>
</blockquote>
<p>where <code>k</code> corresponds to the number of principal components in PCA or bottleneck dimension in AE.</p>
<p>The local adoption is <a href="utils/ae_test.Rmd">ae_test.Rmd</a> which produces <a href="utils/ae_test.html">ae_test.html</a> and <a href="utils/ae_test.pdf">ae_test.pdf</a>.</p>
<p>Additional work will be on variatinoal autoencoder (VAE) as indicated in the references below.</p>
<h2 id="references">REFERENCES</h2>
<p>Bludau I, Frank M, Dörig C. et al. Systematic detection of functional proteoform groups from bottom-up proteomic datasets. Nat Commun 12, 3810 (2021). https://doi.org/10.1038/s41467-021-24030-x</p>
<p>Hofert M, Prasad A, Zhu M (2019). Quasi-Monte Carlo for multivariate distributions viagenerative neural networks. https://arxiv.org/abs/1811.00683, https://CRAN.R-project.org/package=gnn.</p>
<p>Kingma DP, Welling M (2014). Auto-Encoding Variational Bayes. https://arxiv.org/abs/1312.6114, https://keras.rstudio.com/articles/examples/variational_autoencoder.html.</p>
<p>Trivadis SK (2017). Variational autoencoders for anomaly detection. https://rpubs.com/zkajdan/308801.</p>
<h2 id="urls">URLs</h2>
<p>https://github.com/diazale/gt-dimred,
https://github.com/lmcinnes/umap,
https://umap-learn.readthedocs.io/en/latest/.</p>
<hr/>
<h2 id="eth-zurich-u-toronto-team-develops-tool-for-bottom-up-proteomics-proteoform-analysis">ETH Zurich, U Toronto Team Develops Tool for Bottom-Up Proteomics Proteoform Analysis</h2>
<p>Jul 28, 2021 | Adam Bonislawski</p>
<p>NEW YORK – A team led by researchers at ETH Zurich and the University of Toronto has developed a tool that allows for the detection of protein proteoforms in bottom-up proteomics data.</p>
<p>Described in a paper published in June in Nature Communications, the tool, called COPF (COrrelation-based functional ProteoForm) uses peptide correlation analysis to detect differences in proteoform populations across different samples or conditions and could aid researchers as they seek to better understand the role different protein forms play in biology and disease.</p>
<p>The human genome is thought to have around 20,000 protein coding genes, but many of these 20,000 proteins exist in the body in various forms, differentiated by, for instance, post-translational modifications or amino acid substitutions. These different forms are called proteoforms, and it is widely believed that biological processes are guided not just by the proteins present but by what proteoforms are present and in what proportions.</p>
<p>Traditional bottom-up proteomics workflows have provided only limited insight into proteoform populations however, due to the fact that the presence of a particular protein is typically inferred by the detection of just a few of its peptides and that digesting proteins into peptides for mass spec analysis makes it near impossible to link a modified peptide back to a particular proteoform.</p>
<p>Some proteomics researchers have addressed this issue by moving to top-down proteomics, which looks at intact proteins, allowing them to better distinguish between different proteoforms. Top-down proteomics is very technically challenging, however, and is not yet able to analyze proteins with the breadth and depth of bottom-up workflows.</p>
<p>Recently, the development of more reproducible and higher-throughput bottom-up workflows, and particular workflows using data independent-acquisition (DIA) mass spectrometry, have allowed researchers like the Nature Communications authors to apply peptide correlation analysis to the study of proteoforms.</p>
<p>Peptide correlation analysis looks at differences in peptide behavior within and across proteins in bottom-up data.</p>
<p>Researchers have developed a number of approaches for turning peptide measurements into protein data, with most working under the assumption that peptides from the same protein will behave the same way.</p>
<p>In practice, though, that isn't the case. On one hand, there are a number of technical reasons why two peptides from the same protein may not behave the same way. For instance, different digestion efficiencies could lead to some peptides being more abundant than others. Different ionization efficiencies could similarly make one peptide more likely than another to be detected by the mass spec.</p>
<p>The presence of different proteoforms could also play a role. For instance, if a protein is present in both a full-length and truncated form, expression changes observed in the full-length form wouldn't be observable if the peptide being measured wasn't present in the truncated form. Not only would this throw off protein-level quantitation, but it would also mask relative changes in the two protein forms that could be biologically important.</p>
<p>A major challenge to applying this insight has been determining which differences in peptide behavior reflect real technical or biological variation and which are just noise, noted Hannes Röst, research chair in mass spectrometry-based personalized medicine at the University of Toronto and an author on the Nature Communications study.</p>
<p>"In many cases [such variation] was noise," he said. "When you look at traditional shotgun proteomics workflows and data analyses, really the power is not at the peptide-level quantification but at the protein level from the aggregation of multiple peptides. On the peptide level you see a lot of noise, and I think that has prevented us from using this observation that individual peptides could yield a lot of interested information because people really only looked at the protein-level data, because that is what they trusted."</p>
<p>Röst said that the development of targeted protein quantitation approaches like multiple-reaction monitoring (MRM) has demonstrated that individual peptides can be measured with high accuracy, and the development of DIA mass spec approaches has enabled MRM-style peptide quantitation at the proteome scale.</p>
<p>At the same time, improvements in mass spec technology have allowed researchers to collect the kind of large and reproducible datasets required for peptide correlation analysis, he said.</p>
<p>"These are types of experiments we wouldn't have imagined 10 years ago, because for correlation-based approaches to work, you need a relatively large number of samples, and you need low variance," he said. "We are not detecting [proteoforms] that are not changing between different [conditions], we are only detecting those that change. And for this to work we need to have multiple replicates and we need to have different conditions and to be able to measure these peptides with high quantitative accuracy across these conditions."</p>
<p>The COPF tool looks at the intensities of peptides coming from a particular protein across all the samples measured in an experiment and then calculates peptide correlations for all the pairs of peptides coming from that protein and uses hierarchical clustering to divide the peptides into two clusters. It then scores the likelihood that multiple proteoforms of a protein are present by comparing the level of peptide correlation between the clusters to the level of in-cluster variation.</p>
<p>The tool does not identify the specific modifications or variations that distinguish the different proteoforms but rather the peptides that appear to differentiate between the forms of the protein in the different biological contexts investigated. Analyzing a DIA dataset that looked at five different tissue types across eight different mice, COPF identified 63 proteins that exhibited different proteoform groups, including proteins with known tissue-specific splice variants. The researchers also identified proteoforms created by proteolytic and autocatalytic cleavage and phosphorylation, indicating, they wrote, that the tool is "agnostic to the different mechanisms by which proteoforms can be generated inside the cell."</p>
<p>The development of COPF follows the publication last year of a study by researchers at Barts Cancer Institute and the University of Wisconsin-Madison detailing another peptide correlation analysis tool for identifying proteoforms in bottom-up data called PeCorA.</p>
<p>Unlike COPF, which requires proteoforms to differ by two or more peptides, PeCorA can detect proteoforms based on single peptide differences. This makes it a potentially more sensitive tool but also less specific than COPF, Röst said.</p>
<p>More generally, he said that he expected ongoing improvements in mass spec technology would further improve peptide correlation-based approaches like COPF and PeCorA by boosting peptide coverage.</p>
<p>"To kind of cover every possible protein isoform we would need to have complete coverage of every protein, and unfortunately we are currently quite far away from having peptide-level coverage of every protein," he said. "I think that is currently one of the limitations where we are kind of hitting a wall."</p>
<p>Röst added that his lab has begun acquiring data on Bruker's timsTOF Pro platform, "and there we definitely see both an increase in protein coverage and also in the number of peptides we can measure."</p>
<p>"That's why I'm very optimistic that while this is just the first implementation of the method, the data we are producing at this moment is much more complete, and therefore I think it would be even more suitable to our approach than the data we used in the paper," he said.</p></div>
</div>
</div>
<footer class="col-md-12">
<hr/>
<p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
</footer>
<script>
            var base_url = ".",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
<script defer="" src="js/base.js"></script>
<script defer="" src="javascripts/config.js"></script>
<script defer="" src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script defer="" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script defer="" src="search/main.js"></script>
<div aria-hidden="true" aria-labelledby="searchModalLabel" class="modal" id="mkdocs_search_modal" role="dialog" tabindex="-1">
<div class="modal-dialog modal-lg">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="searchModalLabel">Search</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
<form>
<div class="form-group">
<input class="form-control" id="mkdocs-search-query" placeholder="Search..." title="Type search term here" type="search"/>
</div>
</form>
<div id="mkdocs-search-results"></div>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div><div aria-hidden="true" aria-labelledby="keyboardModalLabel" class="modal" id="mkdocs_keyboard_modal" role="dialog" tabindex="-1">
<div class="modal-dialog">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<table class="table">
<thead>
<tr>
<th style="width: 20%;">Keys</th>
<th>Action</th>
</tr>
</thead>
<tbody>
<tr>
<td class="help shortcut"><kbd>?</kbd></td>
<td>Open this help</td>
</tr>
<tr>
<td class="next shortcut"><kbd>n</kbd></td>
<td>Next page</td>
</tr>
<tr>
<td class="prev shortcut"><kbd>p</kbd></td>
<td>Previous page</td>
</tr>
<tr>
<td class="search shortcut"><kbd>s</kbd></td>
<td>Search</td>
</tr>
</tbody>
</table>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div>
<script src="https://unpkg.com/mermaid@8.8.0/dist/mermaid.min.js"></script><script>mermaid.initialize({});</script></body>
</html>
<!--
MkDocs version : 1.1.2
Build Date UTC : 2021-09-18 10:06:38.731051+00:00
-->
