<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="utf-8"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<link href="../img/favicon.ico" rel="shortcut icon"/>
<title>Protein analysis - Caprion</title>
<link href="../css/bootstrap.min.css" rel="stylesheet"/>
<link href="../css/font-awesome.min.css" rel="stylesheet"/>
<link href="../css/base.css" rel="stylesheet"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" rel="stylesheet"/>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/Bash.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/HTML.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/Java.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/JavaScript.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/Perl.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/Python.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/R.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/rust.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/yaml.min.js"></script>
<script>hljs.highlightAll();</script>
</head>
<body>
<div class="navbar fixed-top navbar-expand-lg navbar-dark bg-dark">
<div class="container">
<a class="navbar-brand" href="..">Caprion</a>
<!-- Expander button -->
<button class="navbar-toggler" data-target="#navbar-collapse" data-toggle="collapse" type="button">
<span class="navbar-toggler-icon"></span>
</button>
<!-- Expanded navigation -->
<div class="navbar-collapse collapse" id="navbar-collapse">
<!-- Main navigation -->
<ul class="nav navbar-nav">
<li class="navitem">
<a class="nav-link" href=".."></a>
</li>
<li class="dropdown">
<a class="nav-link dropdown-toggle" data-toggle="dropdown" href="#">Pilot studies <b class="caret"></b></a>
<ul class="dropdown-menu">
<li>
<a class="dropdown-item" href="../pilot/autoencoder/">Autoencoder</a>
</li>
<li>
<a class="dropdown-item" href="../pilot/gwas2/">gwas2</a>
</li>
<li>
<a class="dropdown-item" href="../pilot/">Pilot studies</a>
</li>
</ul>
</li>
<li class="dropdown active">
<a class="nav-link dropdown-toggle" data-toggle="dropdown" href="#">Analysis <b class="caret"></b></a>
<ul class="dropdown-menu">
<li>
<a class="dropdown-item active" href="./">Protein analysis</a>
</li>
<li>
<a class="dropdown-item" href="../peptide_progs/">Peptide analysis</a>
</li>
<li>
<a class="dropdown-item" href="../misc/">Miscellaneous analysis</a>
</li>
</ul>
</li>
<li class="dropdown">
<a class="nav-link dropdown-toggle" data-toggle="dropdown" href="#">Additional information <b class="caret"></b></a>
<ul class="dropdown-menu">
<li>
<a class="dropdown-item" href="https://jinghuazhao.github.io/pQTLdata/reference/caprion.html">Panel information</a>
</li>
<li>
<a class="dropdown-item" href="../pilot/Notes/">Notes</a>
</li>
</ul>
</li>
</ul>
<ul class="nav navbar-nav ml-auto">
<li class="nav-item">
<a class="nav-link" data-target="#mkdocs_search_modal" data-toggle="modal" href="#">
<i class="fa fa-search"></i> Search
                            </a>
</li>
<li class="nav-item">
<a class="nav-link" href="../pilot/" rel="prev">
<i class="fa fa-arrow-left"></i> Previous
                                </a>
</li>
<li class="nav-item">
<a class="nav-link" href="../peptide_progs/" rel="next">
                                    Next <i class="fa fa-arrow-right"></i>
</a>
</li>
<li class="nav-item">
<a class="nav-link" href="https://github.com/jinghuazhao/Caprion"><i class="fa fa-github"></i> GitHub</a>
</li>
</ul>
</div>
</div>
</div>
<div class="container">
<div class="row">
<div class="col-md-3"><div class="navbar-light navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
<div class="navbar-header">
<button class="navbar-toggler collapsed" data-target="#toc-collapse" data-toggle="collapse" title="Table of Contents" type="button">
<span class="fa fa-angle-down"></span>
</button>
</div>
<div class="navbar-collapse collapse card bg-secondary" id="toc-collapse">
<ul class="nav flex-column">
<li class="nav-item" data-level="1"><a class="nav-link" href="#protein-analysis">Protein analysis</a>
<ul class="nav flex-column">
<li class="nav-item" data-level="2"><a class="nav-link" href="#programs1">Programs1</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#1-data-handling-and-pca-projection">1. Data handling and PCA projection</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#2-ggm">2. GGM</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#3-wgcna">3. WGCNA</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#4-pca-and-clustering">4. PCA and clustering</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#5-pgwas3">5. pGWAS3</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#6-meta-analysis">6. Meta-analysis</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#7-variant-identification">7. Variant identification</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#8-hla-imputation5">8. HLA imputation5</a>
<ul class="nav flex-column">
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div></div>
<div class="col-md-9" role="main">
<h1 id="protein-analysis">Protein analysis</h1>
<h2 id="programs1">Programs<sup id="fnref:directory"><a class="footnote-ref" href="#fn:directory">1</a></sup></h2>
<p>Work was done in a named sequence<sup id="fnref:workflow"><a class="footnote-ref" href="#fn:workflow">2</a></sup>.</p>
<pre><code>1_pca_projection.sh
2_ggm.R
3_wgcna.R
4_pca_clustering.R
5_pgwas.sh
6_meta_analysis.sh
7_merge.sh
8_hla.sh
</code></pre>
<div class="mermaid">graph TB
1_pca_projection.sh
2_ggm.R
3_wgcna.R
4_pca_clustering.R --&gt; 5_pgwas.sb --&gt; 6_meta_analysis.sh --&gt; 6_meta_analysis.sb --&gt; 7_merge.sb --&gt; 7_merge.sh --&gt; 0_utils.sb --&gt; 5_pgwas.sb
8_hla.sh
</div>
<p>Chromose X is handled together with autosomes, and the loop from <code>0_utils.sb</code> to <code>5_pgwas.sb</code> is to produce mean-by-genotype/QQ/Manhattan/LocusZoom plots -- the former also implements vep_annotate(), fp_data(), fp() which only requires <code>--array=1</code>. Note also that HetISq() only works inside an interactive R session.</p>
<h2 id="1-data-handling-and-pca-projection">1. Data handling and PCA projection</h2>
<p>The pipeline follows HGI contributions nevertheless only serves for reassurance since the study samples were carefully selected.</p>
<h2 id="2-ggm">2. GGM</h2>
<p>The results are ready to report.</p>
<h2 id="3-wgcna">3. WGCNA</h2>
<p>This can be finalised according to the Science paper.</p>
<h2 id="4-pca-and-clustering">4. PCA and clustering</h2>
<p>The groupings based on unfiltered and DR-filtered proteins can be made on three phases altogether and instead of a classification indicator the first three PCs are used.</p>
<p>The PLINK2 has been used in the pilot studies, but now fastGWA using double transformations of the phenotypic data similar to SCALLOP-Seq analysis. Amazingly, a standard assignment statement inside <code>sapply()</code> would produce <code>.pheno</code>/<code>.mpheno</code> containing the raw data.</p>
<p>The file also includes experiments on normalisation.</p>
<h2 id="5-pgwas3">5. pGWAS<sup id="fnref:pGWAS"><a class="footnote-ref" href="#fn:pGWAS">3</a></sup></h2>
<p>The <code>bgen</code> files were extracted from a list of all samples, the variant IDs of which were for all RSids to allow for multiallelic loci.</p>
<p>The (sb)atch file is extended to produce Q-Q/Manhattan/LocusZoom plots and extreme p values are possible for all plots. Note that LocusZoom 1.4 does not contain 1000Genomes build 37 genotypes for chromosome X and therefore they are supplemented with local files in the required format, namely, <code>locuszoom_1.4/data/1000G/genotypes/2014-10-14/EUR/chrX.[bed, bim, fan]</code>. Now that for the Manhattan plot call for VEP is necessary from <code>0_utils.sb</code>, which also produces mean by genotype/dosage plots.</p>
<h2 id="6-meta-analysis">6. Meta-analysis</h2>
<p>Internally, this follows from the SCALLOP/INF implementation, as designed analogous to a Makefile, i.e.,</p>
<pre><code class="language-bash">6_meta_analysis &lt;task&gt;
</code></pre>
<p>where task = METAL_list, METAL_files, METAL_analysis, respectively in sequence. However, due to time limit on HPC, a call to <code>.sb</code> is made for meta-analysis.</p>
<p>To extract significant variants one may resort to <code>awk 'NR==1||$12&lt;log(1e-6)/log(10)' 1433B-1.tbl</code>, say.</p>
<h2 id="7-variant-identification">7. Variant identification</h2>
<p>An iterative merging scheme is employed; the HLA region is simplified but will be specifically handled. Somewhat paradoxically, forest plots are also obtained here<sup id="fnref:metal"><a class="footnote-ref" href="#fn:metal">4</a></sup>.</p>
<p>A SLURM job is executed, to be followed by collection of results.</p>
<h2 id="8-hla-imputation5">8. HLA imputation<sup id="fnref:HLA"><a class="footnote-ref" href="#fn:HLA">5</a></sup></h2>
<p>This is experimented on several software including HIBAG, CookHLA and SNP2HLA as desribed <a href="https://cambridge-ceu.github.io/csd3/applications/CookHLA.html">here</a>. The whole cohort imputation requests resources exceeding the system limits, so a cardio SLURM job is used instead.</p>
<p>The hped file from CookHLA (or converted from HIBAG) can be used by HATK for association analysis while the advantage of SNP2HLA is that binary ped files are ready for use as usual.</p>
<div class="footnote">
<hr/>
<ol>
<li id="fn:directory">
<p><strong>Directories</strong></p>
<p>This is per Caprion project</p>
<p><code>module load miniconda3/4.5.1
export csd3path=/rds/project/jmmh2/rds-jmmh2-projects/olink_proteomics/scallop/miniconda37
source ${csd3path}/bin/activate</code></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>pgwas</td>
<td>pGWAS</td>
</tr>
<tr>
<td>METAL</td>
<td>Meta-analysis</td>
</tr>
<tr>
<td>HLA</td>
<td>HLA imputation</td>
</tr>
<tr>
<td>peptide_progs</td>
<td>peptide analysis</td>
</tr>
<tr>
<td>reports</td>
<td>Reports</td>
</tr>
</tbody>
</table>
<p>Note that <code>docs.sh</code> copies <code>pilot/utils</code> directory of the pilot studies, so coding under that directory is preferable to avoid overwrite.</p>
<p>To accommodate filteredd results, a suffix "" or "_dr" is applied when appropriate. <a class="footnote-backref" href="#fnref:directory" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:workflow">
<p><strong>workflow</strong> (experimental)</p>
<p><code>module add ceuadmin/snakemake
snakemake -s workflow/rules/cojo.smk -j1
snakemake -s workflow/rules/report.smk -j1
snakemake -s workflow/rules/cojo.smk -c --profile workflow</code></p>
<p>and use <code>--unlock</code> when necessary. <a class="footnote-backref" href="#fnref:workflow" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:pGWAS">
<p><strong>Protein GWAS</strong></p>
<p>GCTA/fastGWA employs MAF&gt;=0.001 (~56%) and geno=0.1 so potentially we can have .bgen files as such to speed up.</p>
<p>GCTA uses headerless phenotype files, generated by <code>5_pgwas.sh</code> which is now unnecessary. <a class="footnote-backref" href="#fnref:pGWAS" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
<li id="fn:metal">
<p><strong>Incomplete gamma function</strong></p>
<p>The .info files for proteins BROX and CT027 could not be obtained from METAL 2020-05-05 with the following error message,</p>
<p><code>FATAL ERROR -
a too large, ITMAX too small in gamma countinued fraction (gcf)</code></p>
<p>An attempt was made to fix this and reported as a fixable issue to METAL GitHub respository (<a href="https://github.com/statgen/METAL/issues/24"><code>https://github.com/statgen/METAL/issues/24</code></a>).
This has enabled Forest plots for the associate pQTLs. <a class="footnote-backref" href="#fnref:metal" title="Jump back to footnote 4 in the text">↩</a></p>
</li>
<li id="fn:HLA">
<p><strong>HLA</strong></p>
<p>A database of 3D structures of Major Histocompatibility Complex, <a href="https://www.histo.fyi/">https://www.histo.fyi/</a></p>
<p>Whole cohort imputation is feasible with a HIBAG reference panel,</p>
<table>
<thead>
<tr>
<th><strong>Locus</strong></th>
<th>A</th>
<th>B</th>
<th>C</th>
<th>DPB1</th>
<th>DQA1</th>
<th>DQB1</th>
<th>DRB1</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>N</strong></td>
<td>1857</td>
<td>2572</td>
<td>1866</td>
<td>1624</td>
<td>1740</td>
<td>1924</td>
<td>2436</td>
</tr>
<tr>
<td>SNPs</td>
<td>891</td>
<td>990</td>
<td>1041</td>
<td>689</td>
<td>948</td>
<td>979</td>
<td>891</td>
</tr>
</tbody>
</table>
<p>while the reference panel is based on the 1000Genomes data (N=503) with SNP2HLA and CookHLA.</p>
<p>It is of note that <code>1000G_REF.EUR.chr6.hg18.29mb-34mb.inT1DGC.markers</code> in the 1000Genomes reference panel has 465 variants with HLA prefix and the partition is as follows,</p>
<table>
<thead>
<tr>
<th><strong>Locus</strong></th>
<th>A</th>
<th>B</th>
<th>C</th>
<th>DPB1</th>
<th>DQA1</th>
<th>DQB1</th>
<th>DRB1</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>HLA_</strong></td>
<td>98</td>
<td>183</td>
<td>69</td>
<td>0</td>
<td>0</td>
<td>33</td>
<td>82</td>
</tr>
</tbody>
</table>
<p>A recent update: PGG.HLA, <a href="https://pog.fudan.edu.cn/pggmhc/">https://pog.fudan.edu.cn/pggmhc/</a>, requires data submission. <a class="footnote-backref" href="#fnref:HLA" title="Jump back to footnote 5 in the text">↩</a></p>
</li>
</ol>
</div></div>
</div>
</div>
<footer class="col-md-12">
<hr/>
<p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
</footer>
<script src="../js/jquery-3.6.0.min.js"></script>
<script src="../js/bootstrap.min.js"></script>
<script>
            var base_url = "..",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
<script src="../js/base.js"></script>
<script src="../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script src="../search/main.js"></script>
<div aria-hidden="true" aria-labelledby="searchModalLabel" class="modal" id="mkdocs_search_modal" role="dialog" tabindex="-1">
<div class="modal-dialog modal-lg">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="searchModalLabel">Search</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<p>From here you can search these documents. Enter your search terms below.</p>
<form>
<div class="form-group">
<input class="form-control" id="mkdocs-search-query" placeholder="Search..." title="Type search term here" type="search"/>
</div>
</form>
<div data-no-results-text="No results found" id="mkdocs-search-results"></div>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div><div aria-hidden="true" aria-labelledby="keyboardModalLabel" class="modal" id="mkdocs_keyboard_modal" role="dialog" tabindex="-1">
<div class="modal-dialog">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<table class="table">
<thead>
<tr>
<th style="width: 20%;">Keys</th>
<th>Action</th>
</tr>
</thead>
<tbody>
<tr>
<td class="help shortcut"><kbd>?</kbd></td>
<td>Open this help</td>
</tr>
<tr>
<td class="next shortcut"><kbd>n</kbd></td>
<td>Next page</td>
</tr>
<tr>
<td class="prev shortcut"><kbd>p</kbd></td>
<td>Previous page</td>
</tr>
<tr>
<td class="search shortcut"><kbd>s</kbd></td>
<td>Search</td>
</tr>
</tbody>
</table>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div>
<script type="module">import mermaid from "https://unpkg.com/mermaid@10.4.0/dist/mermaid.esm.min.mjs";
mermaid.initialize({});</script></body>
</html>
