---
title: "EPCR-PROC in the Caprion pilot study"
author:
- Jing Hua Zhao,..., Adam Buterworth
date: "`r Sys.Date()`"
toc: false
number_sections: true
output:
  bookdown::word_document2: default
bibliography: [packages.bib, ppr.bib]
link-citations: yes
csl: nature-genetics.csl
nocite: '@*'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(engine.opts = list(bash = "-l"))
library(bookdown)
library(knitr)
library(magick)
library(png)
library(tinytex)
library(lattice)
```

```{r, include=FALSE}
knitr::write_bib(file = 'packages.bib')
```

```{bash, echo=FALSE}
module load image-magick-7.0.5-9-gcc-5.4.0-d4lemcc
```

## Annotations
```
   Protein Accession  Gene            Protein.Description
EPCR_HUMAN    Q9UNN8 PROCR Endothelial protein C receptor
PROC_HUMAN    P04070  PROC Vitamin K-dependent protein C

chrom chromStart  chromEnd strand  uniprotName hgncSym         ensGene
chr20   33759957  33764613      +   EPCR_HUMAN   PROCR ENSG00000101000
 chr2  128177518 128186519      +   PROC_HUMAN    PROC ENSG00000115718

https://www.ebi.ac.uk/pdbe/pdbe-kb/proteins/Q9UNN8
https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0010980#s5
https://www.uniprot.org/uniprot/Q9UNN8

Molecule processing

Signal peptidei	1 – 17	Sequence analysis
Chain PRO_0000021191	18 – 238	Endothelial protein C receptor

Topology

Topological domain	18 – 210	Extracellular
Transmembrane	211 – 231	Helical
Topological domain	232 – 238	Cytoplasmic

Sequence

        10         20         30         40         50
MLTTLLPILL LSGWAFCSQD ASDGLQRLHM LQISYFRDPY HVWYQGNASL 
        60         70         80         90        100
GGHLTHVLEG PDTNTTIIQL QPLQEPESWA RTQSGLQSYL LQFHGLVRLV 
       110        120        130        140        150
HQERTLAFPL TIRCFLGCEL PPEGSRAHVF FEVAVNGSSF VSFRPERALW 
       160        170        180        190        200
QADTQVTSGV VTFTLQQLNA YNRTRYELRE FLEDTCVQYV QKHISAENTK 
       210        220        230 
GSQTSRSYTS LVLGVLVGSF IIAGVAVGIF LCTGGRRC

https://www.uniprot.org/blast/uniprot/B20210128A94466D2655679D1FD8953E075198DA802FA1D2?alignment=1
https://www.uniprot.org/peptidesearch/
```

## Files on CSD3

Under ~/rds/projects/Caprion_proteomics/pilot/, 

* EPCR-PROC/ directory contains part of the results in this document
* bgen/ contains association results for phase I as extracted here
* bgen2/ contins association results for phase II
  - EPCR-PROC/ contains results on peptides, All and DR.

## Phase I data

### Q-Q/Manhattan plots

Associaiton statistics were based on invn(ormal) transformation, invnormal(prot)~covariates+genotypes.

```{r qqman1, fig.cap="Figure 1. Q-Q and manhattan plots", fig.show="hold", echo=FALSE}
include_graphics(c("EPCR-PROC/Q9UNN8_invn_turboqq.png","EPCR-PROC/Q9UNN8_invn_turboman.png", 
                   "EPCR-PROC/P04070_invn_turboqq.png","EPCR-PROC/P04070_invn_turboman.png"))
```

### Sentinels

#### EPCR

MarkerName|P|prot|rsid
----------|-|----|----
chr1:198680470|7.7746e-06|Q9UNN8_invn|rs150905683
chr1:242492519|4.52035e-06|Q9UNN8_invn|rs4658474
chr2:40640161|8.47293e-06|Q9UNN8_invn|rs6742589
chr2:108904323|1.79526e-06|Q9UNN8_invn|rs71421322
chr2:129343910|3.38733e-06|Q9UNN8_invn|rs538566482
chr3:3589208|6.97483e-06|Q9UNN8_invn|rs73010413
chr3:63622010|4.92919e-06|Q9UNN8_invn|rs73126516
chr3:115857670|8.18486e-06|Q9UNN8_invn|rs36060209
chr4:26467624|3.08996e-06|Q9UNN8_invn|rs78545240
chr5:117693310|2.17759e-06|Q9UNN8_invn|rs73791119
chr5:178789756|6.07192e-06|Q9UNN8_invn|rs11749206
chr6:10437912|8.87339e-06|Q9UNN8_invn|rs76522721
chr6:31385204|5.65772e-07|Q9UNN8_invn|rs113425623
chr6:150708928|6.90209e-06|Q9UNN8_invn|rs9478711
chr7:46878807|2.58103e-06|Q9UNN8_invn|rs73119156
chr8:113975571|7.68998e-06|Q9UNN8_invn|rs10505197
chr8:141604230|1.29017e-06|Q9UNN8_invn|rs11997990
chr11:8066317|2.82897e-06|Q9UNN8_invn|rs72848407
chr11:82121298|8.22729e-06|Q9UNN8_invn|rs7114139
chr12:119574662|3.5405e-06|Q9UNN8_invn|rs12229183
chr14:45522256|5.26276e-07|Q9UNN8_invn|rs374947851
chr16:54065346|8.6531e-06|Q9UNN8_invn|rs12051239
chr16:86878088|1.07322e-06|Q9UNN8_invn|rs11645358
chr18:65690741|3.02894e-06|Q9UNN8_invn|rs17078446

#### PROC

MarkerName|P|prot|rsid
----------|-|----|----
chr1:14700486|4.41035e-06|P04070_invn|rs1539170
chr1:226790403|4.91707e-06|P04070_invn|rs878149
chr2:85379656|1.95094e-06|P04070_invn|rs2568224
chr3:39025856|2.77768e-06|P04070_invn|rs78768764
chr3:70686181|5.40327e-06|P04070_invn|rs9826254
chr3:106597199|8.58278e-07|P04070_invn|rs56400218
chr6:3251943|7.99169e-06|P04070_invn|rs11963805
chr6:7937069|3.63501e-06|P04070_invn|rs776682351
chr6:145369170|3.93743e-06|P04070_invn|rs117840209
chr7:136707968|5.2286e-06|P04070_invn|rs10246819
chr8:7128386|9.31995e-07|P04070_invn|rs370324853
chr8:99183900|5.71606e-06|P04070_invn|rs2443574
chr9:138730664|2.63476e-06|P04070_invn|rs141376886
chr10:54410153|3.65163e-06|P04070_invn|rs1348351
chr10:70244181|4.25257e-06|P04070_invn|rs35608830
chr11:13498427|8.78973e-07|P04070_invn|rs11022845
chr11:13498427|8.78973e-07|P04070_invn|rs556750517
chr11:23539628|9.03995e-06|P04070_invn|rs7943543
chr12:133801548|9.85288e-06|P04070_invn|rs556045604
chr13:67082089|5.07758e-06|P04070_invn|rs148386003
chr13:81938729|4.35324e-06|P04070_invn|rs117099502
chr13:104582407|6.21026e-06|P04070_invn|rs11842347
chr14:93804153|4.33826e-07|P04070_invn|rs117720114
chr15:65003284|1.67562e-06|P04070_invn|rs144408155
chr18:12739631|9.43079e-06|P04070_invn|rs62096050
chr20:36889971|8.93867e-06|P04070_invn|rs6098808
chr20:36889971|8.93867e-06|P04070_invn|rs767174705
chr21:34516707|4.38852e-06|P04070_invn|rs963950

### Correlation of peptides

```{r heatmap1, fig.cap="Figure 2. Heatmap", fig.height=7, fig.wdith=8, echo=FALSE}
include_graphics("EPCR-PROC/EPCR-PROC-corr.png")
```

## Phase II data

```{r heatmap2, fig.cap="Figure 3. Phase II heatmap", fig.show="hold", fig.height=7, fig.wdith=8, echo=FALSE}
include_graphics(c("EPCR-PROC/EPCR-PROC-phase2-all.png","EPCR-PROC/EPCR-PROC-phase2-corr.png"))
```

```
Correlation:
                                     EPCR_All   EPCR_DR
EPCR_442581804_TLAFPLTIR            0.5113919 0.6174716
EPCR_442582461_LHMLQISYFR           0.2986443 0.4357448
EPCR_442603139_LHM[147.0354]LQISYFR 0.7980766 0.7396526
EPCR_442605396_LHM[147.0354]LQISYFR 0.6105076 0.5631775
EPCR_All                            1.0000000 0.9822826
EPCR_DR                             0.9822826 1.0000000

Linear model:

lm(formula = EPCR_All ~ EPCR_442581804 + EPCR_442582461 + EPCR_442603139 +
    EPCR_442605396, data = EPCR)

Coefficients:
                Estimate Std. Error t value Pr(>|t|)
(Intercept)    1.201e+01  2.372e-02  506.08   <2e-16 ***
EPCR_442581804 7.439e-02  1.914e-03   38.86   <2e-16 ***
EPCR_442582461 1.244e-01  1.772e-03   70.25   <2e-16 ***
EPCR_442603139 4.275e-02  1.422e-04  300.64   <2e-16 ***
EPCR_442605396 3.084e-02  1.459e-04  211.41   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.02638 on 1538 degrees of freedom
Multiple R-squared:  0.9932,    Adjusted R-squared:  0.9931
F-statistic: 5.586e+04 on 4 and 1538 DF,  p-value: < 2.2e-16
```

### Association results

#### Q-Q and Manhattan plots for EPCR_All.

```{r qqman2, fig.cap="Figure 4. Q-Q and manhattan plots for EPCR", fig.show="hold", echo=FALSE}
include_graphics(c("EPCR-PROC/EPCR_All_invn_turboqq.png","EPCR-PROC/EPCR_All_invn_turboman.png"))
```

#### Sentinels

The EPCR and PROC results are shown in the following table.

chr|P|prot|Pos|SNP
---|-|----|----------|---
2|3.07538e-08|EPCR_442581804_TLAFPLTIR|37819409|rs112338777
3|2.22701e-08|PROC_442589612_YLDWIHGHIR|39510601|rs13067113
3|1.11376e-08|EPCR_442582461_LHMLQISYFR|49500024|rs577462830
3|4.58887e-09|EPCR_442603139_LHM.147.0354.LQISYFR|191558143|rs145822522
10|2.56787e-08|EPCR_442603139_LHM.147.0354.LQISYFR|56282688|rs150815432
12|1.86016e-08|PROC_442611348_TFVLNFIK|72065081|rs35805727
17|3.99688e-08|PROC_442597510_TFVLNFIK|30865735|rs2519863
20|2.68308e-08|PROC_442747564_ELNQAGQETLVTGWGYHSSR|33741570|rs376468120
20|4.11194e-08|PROC_DR|33741570|rs376468120
20|2.19713e-08|PROC_442580684_YLDWIHGHIR|33784021|rs11906148
20|3.59909e-11|PROC_442582032_LGEYDLRR|33784021|rs11906148
20|3.0714e-12|PROC_442590967_WELDLDIK|33752110|rs143373163
20|4.72132e-12|PROC_442593037_TFVLN.115.0269.FIK|33741570|rs376468120
20|1.30569e-15|PROC_442580593_TFVLNFIK|33741570|rs376468120
20|6.64232e-12|PROC_442580878_LGEYDLR|33741570|rs376468120
20|6.23775e-13|PROC_442581615_RGDSPWQVVLLDSK|33741570|rs376468120
20|1.66179e-14|PROC_442585077_TFVLN.115.0269.FIK|33773630|rs117249133
20|1.05465e-15|PROC_442605638_TFVLN.115.0269.FIK|33784021|rs11906148
20|2.51841e-13|PROC_442618737_RGDSPWQVVLLDSKK|33784021|rs11906148
20|1.30432e-13|PROC_442652002_DTEDQEDQVDPR|33768650|rs10564067
20|4.41775e-12|PROC_442688824_ELNQAGQETLVTGWGYHSSR|33775200|rs11167260
20|6.29456e-15|PROC_442688952_RGDSPWQVVLLDSK|33745891|rs141474375
20|7.23675e-51|EPCR_442603139_LHM.147.0354.LQISYFR|33754405|rs8119351
20|1.20912e-103|EPCR_442582461_LHMLQISYFR|33746789|rs144917890
20|3.18567e-109|EPCR_All|33769926|rs945961
20|8.01571e-125|EPCR_DR|33769926|rs945961
20|2.57793e-195|EPCR_442581804_TLAFPLTIR|33746789|rs144917890

#### LD information

Four SNPs does not present on the LD reference panel: rs112338777, rs577462830, rs145822522,
rs10564067.

```{bash, echo=FALSE}
cut -d' ' -f2,3 --complement --output-delimiter='|' bgen2/EPCR-PROC/5e-8/caprion-invn.sentinels | \
sed 's/Chrom/chr/;s/_invn//g;s/chr[0-9]*://' > EPCR-PROC/EPCR-PROC.sentinels
```

```{r LD-heatmap, echo=FALSE, fig.cap="Figure 5. LD Heatmap", fig.height=7, fig.wdith=8, message=FALSE, warning=FALSE}
caprion <- Sys.getenv("caprion")
sentinels <- read.table(file.path(caprion,"EPCR-PROC","EPCR-PROC.sentinels"),header=TRUE,sep="|")
ldmatrix <- TwoSampleMR::ld_matrix(with(sentinels,SNP),with_alleles=FALSE)^2
cell_labels <- round(ldmatrix,digits=2)
pheatmap::pheatmap(ldmatrix, display_numbers=cell_labels, fontsize_number=8, cellheight=20)
```

The figure shows clear LD blocks.

## Summary

While there was some indication of possible pQTLs from phase I data,
the phase II results were more convincing. In particular, _All and _DR
data gave the same cis-pQTL for EPCR. Not surprisingly, the significance was also
seen in peptides highly correlated with _All and _DR.
