---
title: "Analysis of phase II Caprion proteomics data"
author:
- Jing Hua Zhao,..., Adam Buterworth
date: "`r Sys.Date()`"
toc: false
number_sections: true
output:
  bookdown::word_document2: default
bibliography: [packages.bib, ppr.bib]
link-citations: yes
csl: nature-genetics.csl
nocite: '@*'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(bookdown)
library(knitr)
library(png)
library(tinytex)
library(lattice)
```

```{r, include=FALSE}
knitr::write_bib(file = 'packages.bib')
```

The documentation provides the reationale and elementary results from the two-phase Caprion proteomics study.

Phase I (study code ZWK) included 196 samples and Phase II (study code ZYQ) contained 1,488 samples.

> Briefly, the method employed immuno-depletion of the most abundant proteins from plasma samples, followed by protein reduction, alkylation and digestion.

> Peptide samples were desalted prior to mass spectrometry analysis.

> A two-stage approach normalization was employed. After normalization, a pattern was identified but could not be explained. The pattern has been left in, with the sponsorâ€™s approval.

## Exploratory analysis

```{r, echo=FALSE}
  load("2020.rda")
# PCA
  ppc <- with(Normalized_All, prcomp(na.omit(Normalized_All[,-1]), rank=50, scale=TRUE))
  pc1pc2 <- with(ppc,x)[,1:2]
# K-means clustering
  km <- kmeans(pc1pc2,2)
# Model-based clustering
  suppressPackageStartupMessages(library(mclust))
  mc <- Mclust(pc1pc2,G=2)
  caprion_mc <- read.csv("ZYQ_PC1_groups_20200703.csv")
  mc_caprion_mc <- cbind(caprion_mc,classification=with(mc,classification))
```
An undisclosed pattern was recovered through PCA.

The clustering based on PC1 and PC2 featured poorly with K-means but perfectly with model-based clustering (**classification**), which 
agreed excellently with the two-class membership provided by Caprion (**pc1_group**) as shown in the following table.

```{r ctable, echo=FALSE}
  with(mc_caprion_mc,table(pc1_group,classification))
```

```{r cluster, fig.cap="Figure 1. Pattern in phase II proteins", fig.height=10, echo=FALSE}
  par(mfrow=c(2,1))
  plot(pc1pc2, col = with(km,cluster), pch=19, cex=0.8)
  points(with(km,centers), col = 1:2, pch = 8, cex = 2)
  title("K-means clustering")
  plot(mc, what=c("classification"))
  title("Model-based clustering")
```

## Association analysis

Association statistics were obtained according to model.

$$
\textrm{invnormal(protein)} \sim \textrm{covariates + genotypes}
$$
For phase II model, the covariates include sample classification according to model-based clustering.

The genotypes were in .bgen format with MAF > 0.01 and the association analysis was furnished with PLINK v2 available on CSD3 through `module load plink/2.00-alpha`.

## Selection of sentinels

The sentinels were obtained via a distance-based iterative merging -- for a given p value cutoff the procedure involves several steps:

1. Obtain a list of genomewide significant variants and their (+/- 1Mb) flanking regions. For HLA region (build 37, chr6:25392021-33392022), only one variant is kept.
2. Identify non-overlapping regions.
3. Select the most significant variant in these regions as sentinels; in cases there are ties, keep only one of them.

Step 2 is implemented using `bedtools`. Step 3 can be facilitated with R:

```r
  library(dplyr)
  sentinels <- merged_results %>%
               group_by(trait,Chrom,Start,End) %>%
               slice(which.min(P))
```

## Phase I and Phase II correlation

Genome-wide p values at both phases were used.

To improve the property of statistics involving p values, they are converted to their normal(0, 1) deviates ($z$) for calculation of correlations.

Without loss of generality, let $z$>0 (or $z$=$-z$ otherwise) and for a two-sided p value, we have 

$$
p=\Phi(-z)+(1-\Phi(z)) = 2\Phi(-z)
$$

due to symmetry, so $z=\Phi^{-1}(p/2)$ where $\Phi(.)$ is the cumulative normal distribution function.

```{r function, fig.cap="Figure 2. Normal(0,1) distribution", fig.width=6, fig.height=4, echo=FALSE}
plotp <- function(mean=0, sd=1, z=1.96)
{
  lb <- -z; ub <- z
  x <- seq(-5,5,length=100)*sd + mean
  hx <- dnorm(x,mean,sd)
  plot(x, hx, type="n", xlab="z", ylab="", main="Normal Distribution", axes=FALSE)
  lines(x, hx)
  i <- x <= lb
  polygon(c(-5,x[i],lb), c(0,hx[i],0), col="red")
  i <- x >= ub
  polygon(c(ub,x[i],5), c(0,hx[i],0), col="red")
  area <- pnorm(lb, mean, sd) + 1 - pnorm(ub, mean, sd)
  result <- paste("P(z<=",lb,"| z>=",ub,") =", signif(area, digits=3))
  mtext(result,3)
  axis(1, at=c(-z, 0, z), pos=0)
}

plotp()
```

```{r normal, echo=FALSE, eval=FALSE, fig.cap="Figure 2. Normal(0,1) distribution", fig.width=6, fig.height=4}
z0 <- 1.96
z <- 5
ea <- seq(-z, z, length = 10000)
eb <- dnorm(ea, 0, 1)
        xyplot(eb ~ ea,
               type = "l",
               panel = function(x,y, ...)
               {
                   panel.xyplot(x,y, ...)
                   panel.abline(v = 0, lty = 2)
                   xx <- c(-z, x[x>=-z & x<=-z0], -z0)
                   yy <- c(0, y[x>=-z & x<=-z0], 0)
                   panel.polygon(xx,yy, ..., col='red')
                   xx <- c(z0, x[x>=z0 & x<=z], z)
                   yy <- c(0, y[x>=z0 & x<=z], 0)
                   panel.polygon(xx,yy, ..., col='red')
               },
               xlab="z",
               ylab=expression(1/sqrt(2*pi) * exp(-z^2/2)))
```

We therefore have the R function `z <- function(p) -qnorm(p/2)`.

## Miami plots

This visually contrasts two vectors of p values emanating from the same chromosomal positions, and an example on P22626-ROA2 \@ref(fig:P22626_ROA2) is given below.

The plot was produced with miamiplot() function in R/gap.

```{r P22626_ROA2, fig.cap="Figure 3. P22626-ROA2 Miami plot", fig.height=4.5, fig.wdith=4.5, echo=FALSE}
include_graphics("miamiplot/24-03-2021/P22626-ROA2-phase1-phase2.png")
```

## Overlap of sentinels

Our focus here is on $10^{-5}$ and $5 \times 10^{-8}$.

# Files at CSD3

## Data

/rds/project/jmmh2/rds-jmmh2-projects/Caprion_proteomics/

## Analysis

pilot/ subdirectory of data.

Item           | I                   | II
---------------|---------------------|-------------------
Data           | data                | data2
PLINK2 results | bgen (1e-5, 5e-8)   | bgen2 (1e-5, 5e-8)

where sentinels from specific p value cutoff are in their own named directories, i.e., 1e-5, 5e-8.

# A summary

Item      |I/protein| II/protein
----------|---------|-----------
N         | 196     | 1488
----------|---------|-----------
Sentinels |
----------|---------|-----------
1e-5      |25766/987| 16298/982
----------|---------|-----------
5e-8      |226/203  | 192/178
----------|---------|-----------

Here phase II results were based on all data (_All, normalised log-intensities for every sample).

# A case study

The EPCR-PROC case study was reported earlier.

# Reflections

It is somewhat counterintuitive not to see more significant results from phase II although the sample size is seven fold larger. It is unclear if this is due to a systematic bias in the measurement.

It is possible to work on the piptides and/or also only those above limit of detection (_DR, normalised log-intensities for every sample with detection rate >10%).

# References

<div id="refs"></div>

# (APPENDIX) Appendix {-}

# Correlation between p values and their difference

The idea is best illustrated with a toy example involving two vectors of p values (`p` and `pr`) below

```{r echo=TRUE}
# z=-1.96 correspond to a two-sided p=0.05
z <- function(p) -qnorm(p/2)
z(0.05)
p <- 1:10/10; p
pr <- 1:10/100; pr
cor(z(p),z(pr))
wilcox.test(p,pr,alternative="less",paired=TRUE)
```
where the correlation is obtained according to [0,1] and a nonparametric test of difference conducted. Note in particular the way the statistical significance was reported, for `pr` was generated an order of magnitude smaller than `p`.

# Information regarding the Miami plot

```{r echo=FALSE}
img_path <- "miamiplot/24-03-2021/P22626-ROA2-phase1-phase2.png"
img <- readPNG(img_path, native = TRUE, info = TRUE)
attr(img, "info")
```
