---
title: "Caprion proteomics phase I/II studies"
author:
- Jing Hua Zhao,..., Adam Buterworth
date: "`r Sys.Date()`"
toc: true
number_sections: true
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(png)
```

The documentation considers the reationale and elementary results the two-phase Caprion proteomics study.

## Statistical models

Association statistics were obtained according to the following model.

> invnormal(protein) ~ covariates + genotypes

The genotypes were in .bgen format with MAF > 0.01 and the association analysis was furnished with PLINK v2 available on CSD3 through `module load plink/2.00-alpha`.

## Phase I and Phase II correlation

Genome-wide p values at both phases were used.

To improve the property of statistics involving p values, they are converted to their normal(0, 1) deviates ($z$) for calculation of correlations. Without loss of generality, let $z$>0 and for a two-sided p value, we have $p=\Phi(-z)+(1-\Phi(z)) = 2\Phi(-z)$ due to symmetry, so $z=\Phi^{-1}(p/2)$ where $\Phi(.)$ is the cumulative distribution function.

We therefore have the R function `z <- function(p) -qnorm(p/2)`.

## Miami plots

This visually contrasts two vectors of p values emanating from the same chromosomal positions, and an example is given below.

```{r echo=FALSE, fig.cap="P22626-ROA2 Miami plot"}
include_graphics("miamiplot/24-03-2021/P22626-ROA2-phase1-phase2.png")
```

The plot was produced with miamiplot() function in R/gap.

## Overlap of sentinels

The sentinels were obtained via a distance-based iterative merging -- for a given p value cutoff the procedure involves several steps:

1. Obtain a list of genomewide significant variants and their (+/- 1Mb) flanking regions. For HLA region (chr6:25392021-33392022), only one variant is kept.
2. Identify non-overlapping regions.
3. Select the most significant variant in these regions as sentinels; in cases there are ties, keep only one of them.

Step 2 is implemented using `bedtools`. Step 3 is facilitaed with an R:

```r
  library(dplyr)
  sentinels <- merged_results %>% group_by(trait,Chrom,Start,End) %>% slice(which.min(P))
```

Our focus here is on $10^{-5}$ and $5 \times 10^{-8}$.

# Files at CSD3

## Data

/rds/project/jmmh2/rds-jmmh2-projects/Caprion_proteomics/

## Analysis

pilot/ subdirectory of data.

Item           | I                   | II
---------------|---------------------|-------------------
Data           | data                | data2
PLINK2 results | bgen (1e-5, 5e-8)   | bgen2 (1e-5, 5e-8)

where sentinels from specific p value cutoff are in their own named directories, i.e., 1e-5, 5e-8.

# A summary

Item      |I/protein| II/protein
----------|---------|-----------
N         | 196     | 1488
----------|---------|-----------
Sentinels |
----------|---------|-----------
1e-5      |25766/987| 16298/982
----------|---------|-----------
5e-8      |226/203  | 192/178
----------|---------|-----------

Here phase II results were based on all data (_All).

# A case study

The EPCR-PROC case study was reported earlier.

# Possible work

It is possible to work on the piptides and also only those above limit of detection (_DR).

# Appendix: A toy example

The idea is best illustrated with a toy example involving two vectors of p values (`p` and `pr`) below

```{r echo=TRUE}
# z=-1.96 correspond to a two-sided p=0.05
z <- function(p) -qnorm(p/2)
z(0.05)
p <- 1:10/10; p
pr <- 1:10/100; pr
cor(z(p),z(pr))
wilcox.test(p,pr,alternative="less",paired=TRUE)
```
where the correlation is obtained according to $n(0,1)$ and a nonparametric test of difference conducted. Note in particular the way the statistical significance was reported, for `pr` was generated an order of magnitude smaller than `p`.

Information regarding the Miami plot above.

```{r echo=FALSE}
img_path <- "miamiplot/24-03-2021/P22626-ROA2-phase1-phase2.png"
img <- readPNG(img_path, native = TRUE, info = TRUE)
attr(img, "info")
```
